utils/prof_utils.py

```diff
@@ -0,0 +1,819 @@
+        if val==0.:
+            on_pulse_prof.append(profile[i])
+            on_pulse_bins.append(i)
+        else:
+            on_pulse_prof.append(0)
+
+    #regularization with Stickel - smoothing
+    x = np.linspace(0,len(profile)-1, len(profile), dtype=int)
+    y = np.array(on_pulse_prof)
+    data = np.column_stack((x, y))
+    stkl = Stickel(data)
+    stkl.smooth_y(reg_param)
+    on_pulse_prof = stkl.yhat
+
+    #subract noise mean and normalize profile
+    noise_mean = np.nanmean(clipped_prof)
+    on_pulse_prof = np.array(on_pulse_prof) - noise_mean
+    on_pulse_prof = on_pulse_prof/np.max(on_pulse_prof)
+    #reset noise to 0
+    for i, val in enumerate(on_pulse_prof):
+        if val<0:
+            on_pulse_prof[i]=0
+
+    return on_pulse_prof, noise_mean, noise_std
+
+def multi_gauss(x, *params):
+    y = np.zeros_like(x)
+    for i in range(0, len(params), 3):
+        ctr = params[i]
+        amp = params[i+1]
+        wid = params[i+2]
+        y = y + amp * np.exp( -((x - ctr)/wid)**2)
+    return y
+
+def fit_gaussian(profile, max_N=6, chi_threshold=0, min_comp_len=0, plot_name=None):
+    """
+    Fits multiple gaussian components to a pulse profile and finds the best number to use for a fit.
+    Will always fit at least one gaussian per profile component.
+    Profile components are defined by find_components().
+    Each gaussian is defined by the following: y = amp * np.exp( -((x - ctr)/wid)**2)
+
+    Parameters:
+    -----------
+    profile: list
+        A list containing the profile data
+    max_N: int
+        OPTIONAL - The maximum number of gaussain components to attempt to fit. Default: 6
+    chi_threshold: float
+        OPTIONAL - The script will stop trying new fits when the reduced chi-squared is within this amount to unity. Default: 0
+    plot_name: string
+        OPTIONAL - If not none, will make a plot of the best fit with this name. Default: None
+
+    Returns:
+    --------
+    [fit, chisq, popt, pcov]: list
+        fit: list
+            The data containing the multi-component gaussian fit to the input profile
+        chisq: float
+            The reduced chi-sqaured value of the fit
+        popt: list
+            A list of floats where each 3 numbers describes a single gaussain and are 'ctr', 'amp' and 'wid' respectively
+        pcov: numpy matrix
+            The covariance matrix generated by the curve_fit function
+    """
+    #chi sqaured evaluation
+    def chsq(observed_values, expected_values, err):
+        test_statistic=0
+        for observed, expected in zip(observed_values, expected_values):
+            test_statistic+=((float(observed)-float(expected))/float(err))**2
+        return test_statistic
+
+    #Take noise mean and normalize the profile
+    _, clipped = sigmaClip(profile)
+    y = np.array(profile) - np.nanmean(np.array(clipped))
+    max_y = max(y)
+    len_y = len(y)
+    y = np.array(y)/max_y
+    noise_mean = np.nanmean(np.array(clipped)/max_y)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'noise_mean'](https://app.codacy.com/app/CIRA-Pulsars-and-Transients-Group/vcstools/pullRequest?prid=4645369)